### PR TITLE
Add support for looking up client secrets via Vault SPI

### LINKED ISSUE #13102: Add support for specifying `client.secret` using vault
### Description

Currently, only LDAP/SMTP/IDP passwords support providing value using the vault. This issue should add support also for client secrets.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by tsaarni**
This PR adds support for resolving client secrets via the Vault SPI, enabling retrieval from secure storage instead of storing them in plaintext in the database. While the Vault SPI has some usability limitations, this approach is less invasive than alternatives like [hashing](https://github.com/keycloak/keycloak/discussions/8455) or introducing new  [encrypt/decrypt SPI](https://github.com/keycloak/keycloak-community/pull/355). It also preserves the plaintext access to the secret, which is required by existing use cases.

Notes:

(1) At the moment, the admin console only supports generating new random client secrets. There is no way for administrators to specify a vault reference value. However, the Admin REST API does support setting the client secret explicitly, including reference values. To fully support this feature, the admin console will need to be extended to allow entering reference values as well. Note: If a new random client secret is generated via the admin console, it will overwrite the existing reference.

(2) The [client secret rotation feature](https://www.keycloak.org/docs/latest/server_admin/index.html#_secret_rotation) introduces additional complexity. Triggering rotation will overwrite any reference-based secret with a newly generated random secret.

(3) Since the client secret may be accessed frequently, using a remote Vault SPI implementation with network communication can introduce latency. To avoid repeated lookups, Vault SPI implementation may choose to add caching.

TODO
- [x] Add test cases for the new Vault-based client secret resolution.
- [x] Extend the admin console to support setting client secret references.
- [x] Add documentation.
- [x] Update documentation with new screenshot.

Fixes #13102

### CODE REVIEW

**Comment by tsaarni** [comment]
Hi @mhajas, I noticed that you had opened an issue requesting this feature a few years ago. I’m interested in contributing it and would be glad to continue working on it. Before proceeding, I wanted to check if this is still a feature you'd consider including in Keycloak?

**Comment by mhajas** [comment]
Thank you for the PR @tsaarni. I would say it still makes sense to include this since it is highly upvoted issue. I would be happy to help or find someone else who would be willing to help. 

I will have a look at the changes later this week. Let me know if you need something else at this stage.

**Comment by tsaarni** [comment]
I've now marked the PR as ready for reviews.

I still have a couple of basic question regarding the UI changes.

**(1)** I didn't add an option to set the client secret during client creation. As a result, the client is always created with randomly generated client secret (as before), even if administrator already knows they want to use a vault reference. In this case, they'll need to first create the client with a random secret, then go to the "Credentials" tab to configure the vault reference.

Do you think this is acceptable from a user experience perspective?

**(2)** In he "Credentials" tab, there was previously a separator line between authenticator selector and the client secret field:

client-credentials-before

I've removed that separator and moved the Client Secret field above the "Save" button, to make i more consistent with the layout used for the other authenticators:

client-credentials

I haven't updated the documentation with a new UI screenshot yet, as I'd like to agree on the approach first.

**(3)** Side effect of this change is that administrators can now set any specific client secret directly through the UI, which wasn’t previously possible. This capability has existed via the REST API, but not in the UI until now. Hope this is acceptable.

<del>**(4)** In other Vault SPI use cases, the secret stored in the vault is not accessible through the UI. However, for OIDC clients, exporting the client configuration from the UI previously included the actual secret. The configuration comes from the client representation, which contains the vault reference instead of the secret itself. I haven’t found a workaround yet, and I’m uncertain whether the secret should be resolved during export.</del> This was just my misunderstanding. There is separate endpoint for downloading client config so I changed that to resolve the vault secret. Now the downloaded client configuration file has the secret itself instead of vault reference.

Lastly, I haven’t used React Hook Form before, but I tried to follow the existing patterns used elsewhere in the code.

**Comment by tsaarni** [comment]
I've updated the PR and addressed the (4) [client config download issue](https://github.com/keycloak/keycloak/pull/39650#issuecomment-2962693402) I had. I’m looking forward to receive reviews and happy to address any comments to help get this merged.

**Comment by tsaarni** [comment]
I've rebased the PR.
Since @mhajas is no longer involved with Keycloak, I was wondering if another maintainer could review it. Thank You!

File: js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx [hunk 1]
```diff
@@ -47,7 +47,11 @@ const SecretInput = ({
       <SplitItem isFilled>
         <InputGroup>
           <InputGroupItem isFill>
-            <PasswordInput id={id} value={secret} readOnly />
+            <Controller
+              name="secret"
+              control={form.control}
+              render={({ field }) => <PasswordInput id={id} {...field} />}
+            />
```
Comment (rmartinc) [inline js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: With this change we are allowing editing any secret (even for view only users that cannot submit). I think that this was initially read-only because the idea is only allow the generate button. This way the passwords are secure/complex. Maybe we should maintain it read-only initially and allow editing when changed to visible (only for admins that can edit). Not sure about this. I thought about showing a message when toggle to visible, commenting about complexity and so on and so forth... But maybe it's too much.

But I would at least present it read-only and just enable editing when toggled to visible if the admin has permissions for edit (if not it remains read-only).

Change the documentation accordingly if you think my proposal is better.

@edewit @tsaarni WDYT? Any other idea?

File: server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java [hunk 2]
```diff
@@ -54,6 +55,16 @@ public interface ClientAuthenticatorFactory extends ProviderFactory<ClientAuthen
      */
     Map<String, Object> getAdapterConfiguration(ClientModel client);
 
+    /**
+     * Get configuration, which needs to be used for adapter ( keycloak.json ) of particular client. Some implementations
+     * may return just template and user needs to edit the values according to his environment (For example fill the location of keystore file)
+     *
+     * @return
+     */
+    default Map<String, Object> getAdapterConfiguration(KeycloakSession session, ClientModel client) {
```
Comment (rmartinc) [inline server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java]: This is the `server-spi-private`, so I think that we can just change the `getAdapterConfiguration` signature. Do not add a new method, just change the existing one. You will need to change the `ClientIdAndSecretAuthenticator`, `FederatedJWTClientAuthenticator` and so on and so forth.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java [hunk 3]
```diff
@@ -30,16 +31,19 @@
...
-    public static OIDCClientSecretConfigWrapper fromClientModel(ClientModel client) {
-        return new OIDCClientSecretConfigWrapper(client, null);
+    public static OIDCClientSecretConfigWrapper fromClientModel(ClientModel client, KeycloakSession session) {
+        return new OIDCClientSecretConfigWrapper(client, null, session);
     }
 
-    public static OIDCClientSecretConfigWrapper fromClientRepresentation(ClientRepresentation clientRep) {
-        return new OIDCClientSecretConfigWrapper(null, clientRep);
+    public static OIDCClientSecretConfigWrapper fromClientRepresentation(ClientRepresentation clientRep, KeycloakSession session) {
+        return new OIDCClientSecretConfigWrapper(null, clientRep, session);
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: I think that I prefer to not add the session in the constructor and just pass the sessions in the required methods. They are just two. WDYT?

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java [hunk 4]
```diff
@@ -102,7 +106,8 @@ public boolean hasRotatedSecret() {
     }
 
     public String getClientRotatedSecret() {
-        return getAttribute(CLIENT_ROTATED_SECRET);
+        String secret = getAttribute(CLIENT_ROTATED_SECRET);
+        return session == null ? getAttribute(CLIENT_ROTATED_SECRET) : session.vault().getStringSecret(secret).get().orElse(secret);
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java:107]: This method can just receive the session as parameter. You would need to change the callers but they are not too many (for the test just pass null).

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java [hunk 5]
```diff
@@ -180,6 +185,19 @@ public boolean isClientRotatedSecretExpired() {
         return true;
     }
 
+    public String getClientPrimarySecret() {
+        String secret = clientModel.getSecret();
+        return session == null ? secret : session.vault().getStringSecret(secret).get().orElse(secret);
+    }
+
+    public boolean validatePrimarySecret(String secret) {
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: Same for this method, if you pass the session as parameter we can remove the session at creation time. I would just call it `validateSecret` and I would remove the `getClientPrimarySecret` method. I would do something like:

```java
    public boolean validateSecret(KeycloakSession session, String secret) {
        if (isClientSecretExpired()) {
            return false;
        }

        ClientModel wrapper = new ClientModelLazyDelegate(() -> clientModel) {
            @Override
            public String getSecret() {
                final String secret = clientModel.getSecret();
                final String result = session.vault().getStringSecret(secret).get().orElse(secret);
                return result;
            }

            @Override
            public boolean validateSecret(String secret) {
                return MessageDigest.isEqual(secret.getBytes(), getSecret().getBytes());
            }
        };

        return wrapper.validateSecret(secret);
    }
```

You should change the authenticator that calls this one with the new signature.

**Review by rmartinc — CHANGES_REQUESTED** [review]
Thanks @tsaarni for the PR!

I have added some comments. I worked with this [branch](https://github.com/rmartinc/keycloak/tree/refs/heads/pr-39650) for the review. So you can take whatever you need from that extra test commit i played with.

I have doubts about the admin console change. Asking @edewit if he can have a better idea.

Probably, when the PR is ready will need a note in the release notes about the change. But we can talk about that later.

File: js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx (hunk 1 above)
Comment (tsaarni) [reply js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: >  I think that this was initially read-only because the idea is only allow the generate button. This way the passwords are secure/complex.

Likely true, though the client secrets were possible to change via REST API anyways. 

> But I would at least present it read-only and just enable editing when toggled to visible if the admin has permissions for edit (if not it remains read-only).

Thanks! I had missed the `readOnly={!isManager}`.  Regarding being editable only after toggling visibility, I suppose `<PasswordInput>` component remains editable by design even when password is not revealed?

**Comment by tsaarni** [comment]
Big thanks for review @rmartinc! I incorporated everything from your commit, thanks for making it available :)

File: server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java (hunk 2 above)
Comment (tsaarni) [reply server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java]: Done. I hope people have not implemented this, regardless of it being private :)

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 3 above)
Comment (tsaarni) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: Done.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 5 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: :+1: You can resolve this one.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 4 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java:107]: :+1: You can resolve this one.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 3 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: :+1: You can resolve this one.

File: server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java (hunk 2 above)
Comment (rmartinc) [reply server-spi-private/src/main/java/org/keycloak/authentication/ClientAuthenticatorFactory.java]: :+1: Yes, but the old one should also be deprecated, so I prefer to just change the method. It's private, and the change is also very simple.

You can resolve this one.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 4 above)
Comment (tsaarni) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java:107]: Done.

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java (hunk 5 above)
Comment (tsaarni) [reply services/src/main/java/org/keycloak/protocol/oidc/OIDCClientSecretConfigWrapper.java]: Done.

File: js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx (hunk 1 above)
Comment (rmartinc) [reply js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: I was thinking about something similar to https://github.com/rmartinc/keycloak/commit/d59fa4601d2274a90da503c93b8cfdb0912f1057. I mean, by default is read-only, and only when revealed is also editable. But I'm not sure if I'm just overthinking and  complicating the UI. We can go even farther and can show a warning or similar before make it editable with this idea.

@edewit @mposolda WDYT? Do we always allow modifying the secret or just when revealing it? I was trying to make the generate button the easiest path except if you really want to type the secret (because you want a vault or any other reason).

Comment (edewit) [reply js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: I think it's fine to leave it like this that when you want to edit it you can only do that when it's viewable, otherwise we would have to have 2 fields to confirm the new password. One thing to note is instead of using `<Control` it would be better to use 
```
            <PasswordControl
              name="secret"
              isDisabled={!isManager}
            />
```

Comment (rmartinc) [reply js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: Thanks @edewit! I think this is using directly the `PasswordInput` instead of the `PasswordControl` because there are more elements in the row (the copy and regenerate buttons and more things for the client secret rotation). And you need to change the `PasswordControl` to add everything inside it.

Then @tsaarni, if Erik thinks that it is OK like it is now, just change the the read-only for disabled (it's the same and I think that it works like the rest of fields) and that's all:

```js
            <Controller
              name="secret"
              control={form.control}
              render={({ field }) => (
                <PasswordInput id={id} {...field} isDisabled={!isManager} />
              )}
            />
```

Comment (tsaarni) [reply js/apps/admin-ui/src/clients/credentials/ClientSecret.tsx:56]: Thanks! I've changed from `readOnly` to `isDisabled`.

**Review by rmartinc — APPROVED** [review]
Thanks @tsaarni! LGTM!

@mposolda I think this is OK. The only missing point is that probably we need to add a little release note for 26.6 about this. Don't know if you want to add it in this PR or later.

File: services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java [hunk 6]
```diff
@@ -142,7 +144,7 @@ public List<ProviderConfigProperty> getConfigPropertiesPerClient() {
     }
 
     @Override
-    public Map<String, Object> getAdapterConfiguration(ClientModel client) {
+    public Map<String, Object> getAdapterConfiguration(KeycloakSession session, ClientModel client) {
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: Minor: Should not this method use the same like `ClientIdAndSecretAuthenticator.getAdapterConfiguration` to actually obtain secret for adapters?

File: docs/documentation/server_admin/topics/clients/oidc/con-confidential-client-credentials.adoc [hunk 7]
```diff
@@ -12,7 +12,8 @@ The *Client Authenticator* drop-down list specifies the type of credential to us
 
 *Client ID and Secret*
 
-This choice is the default setting. The secret is automatically generated. Click *Regenerate* to recreate the secret if necessary.
+This choice is the default setting. A random secret is generated automatically, but you can override it with a specific value if needed. The client secret can also reference a value stored in an external vault. Click *Regenerate* to recreate random secret if necessary.
```
Comment (mposolda) [inline docs/documentation/server_admin/topics/clients/oidc/con-confidential-client-credentials.adoc]: Minor: Is it possible that text "external vault" adds link to the Vault chapter docs?

**Review by mposolda — COMMENTED** [review]
@tsaarni @rmartinc @edewit Thanks for the PR and review!

Added few minor comments inline.

Besides that, I am missing a bit a "Revert" button on the page with client credentials. This button was not present here before this PR as well (and is also missing for "X509" credential where it would also makes sense). But with disabled "Regenerate" button after doing some changes in client-secret, I am missing it a bit more now.

For example: Assume I want to edit client-secret by hand, so I would start typing something in the client-secret field. But then I decide I rather prefer to "Regenerate" secret instead of manually updating it. I cannot do that very easily (unless I delete it to previous value OR unless I go back from the client entirely OR unless I am first "Save" it and then "Regenerate" as next step, which all are not so great options from UX perspective IMO).

@tsaarni @rmartinc @edewit Curious for the feedback on this. If you think "Revert" button not needed, I am OK with accepting this without "revert" button (As I can see that "Revert" button can imply that "Regenerating" client secret could be also reverted, but it actually could not as REST request for regenerate was already saved and previous secret already lost. So "revert" button should be rather present/enabled just if `Save` button is enabled as well).

File: services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java (hunk 6 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: Yep! I missed this one. It should also use the `vault` to get the secret.

**Comment by rmartinc** [comment]
@mposolda @tsaarni I'm OK adding the revert button.

File: services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java (hunk 6 above)
Comment (tsaarni) [reply services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: I only planned to add vault support for the client secret authenticator but I now tested the JWT client secret authenticator as well and it's not working with vault references. Not sure what's needed to implement vault support for that as well.

File: docs/documentation/server_admin/topics/clients/oidc/con-confidential-client-credentials.adoc (hunk 7 above)
Comment (tsaarni) [reply docs/documentation/server_admin/topics/clients/oidc/con-confidential-client-credentials.adoc]: Done.

File: services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java (hunk 6 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: OK, but then this should be at least managed as follow-up. The JWTClientSecretAuthentication is just an authenticator that uses the secret as the HMAC to sign... We would need also tests for this. @mposolda WDYT?

Comment (mposolda) [reply services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: Ah, ok. It would be good to update that one as well for consistency. It seems this authenticator uses key, which is created from client-secret and secret is obtained in `ClientMacSignatureVerifierContext` (not 100% sure it is this class. You can likely doublecheck).

Comment (tsaarni) [reply services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java:147]: @mposolda @rmartinc I've now added support for JWT client secret authenticator + test case.

**Review by mposolda — APPROVED** [review]
@tsaarni @rmartinc @edewit Thanks! Revert button can be possibly nice, but likely not strictly needed. Could be added as a follow-up somewhen if people ask for it ( @tsaarni If you want to contribute it in a follow-up PR, it would be welcome :-) )

**Comment by tsaarni** [comment]
I'm working on the button right now, so I will come back with a button bit later today (or admitting defeat and moving it to someone else to follow-up ;) )
